### PR TITLE
feat(artifacts):  prototype for models api

### PR DESCRIPTION
**Pull request by ibindlish**
Fixes
-----
- Fixes [WB-15814](https://wandb.atlassian.net/browse/WB-15814)
- Fixes #NNNN

Description
-----------
Prototype for model APIs:
- `log_model()` (note: `max_objects` param while adding reference is not exposed on log_artifact, hence not exposed for log_model)
- `use_model()` - returns path to directory where we download the artifact files. If there's only 1 file, then we return the path to that file.
- `link_model()`

Re. [Notion doc](https://www.notion.so/wandbai/Model-logging-API-5d746ea5daa4486980c968a19dd3103a)

### <samp>🤖 Generated by Copilot at 78991d8</samp>

Add new functions to `wandb/apis/models.py` for logging, using, and linking model artifacts. These functions simplify the workflow of working with model artifacts in wandb.

Testing
-------

```
tmpdir = "../../test-examples"

log_model(tmpdir, model_name="test-log-model")
data_dir = use_model("testing-log-model:latest")
link_model(
    local_path=tmpdir,
    linked_model_name="testing-link-model-api",
    aliases=["testing-link"],
)
```
- [x] Small artifacts
- [x] Reference artifacts (** For references, tested using s3.)
- [x] Large artifacts
- [x] Large reference artifacts (tested by hackily adding `max_objects=10**10` in wandb _log_artifact() for testing purposes; but from a user pov, till we set the default for max_objects as `None` they won't be able to log large artifact references)

Checklist
-------
- [x] Include reference to internal ticket "Fixes WB-NNNNN" and/or GitHub issue "Fixes #NNNN" (if applicable)
- [x] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/main/CONTRIBUTING.md#conventional-commits)

### <samp>🤖 Generated by Copilot at 78991d8</samp>

> _`log_model`, `use_model`_
> _New functions for wandb users_
> _Cutting edge in spring_

[WB-15814]: https://wandb.atlassian.net/browse/WB-15814?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: wandb/sdk/wandb_run.py [hunk 1]
```diff
@@ -3072,6 +3075,93 @@ def _prepare_artifact(
...
+        aliases: Optional[List[str]] = None,
+    ) -> None:
+        """Declare a model artifact as an output of a run.
+
+        Arguments:
+            local_path: (str) A path to the contents of this model,
+                can be in the following forms:
+                    - `/local/directory`
+                    - `/local/directory/file.txt`
+                    - `s3://bucket/path`
```
Comment (szymon-piechowicz-wandb) [inline wandb/sdk/wandb_run.py:3091]: Nit: this isn't local, so `local_path` should just be `path`.

File: wandb/sdk/wandb_run.py [hunk 2]
```diff
@@ -3072,6 +3075,93 @@ def _prepare_artifact(
...
+            model_name: (str, optional) An artifact name. May be prefixed with entity/project.
+                Valid names can be in the following forms:
+                    - name:version
+                    - name:alias
+                    - digest.
```
Comment (szymon-piechowicz-wandb) [inline wandb/sdk/wandb_run.py:3096]: Nit: this is incorrect. The name gets passed to the `Artifact` constructor.

File: wandb/sdk/wandb_run.py [hunk 3]
```diff
@@ -3072,6 +3075,93 @@ def _prepare_artifact(
...
+            aliases: (list, optional) Aliases to apply to this artifact,
+                    defaults to `["latest"]`
+
+        Returns:
+            None
+        """
+        self._log_artifact(
+            artifact_or_path=local_path, name=model_name, type="model", aliases=aliases
+        )
+        return None
```
Comment (szymon-piechowicz-wandb) [inline wandb/sdk/wandb_run.py:3108]: Nit: remove, this is not needed.

File: wandb/sdk/wandb_run.py [hunk 4]
```diff
@@ -3072,6 +3075,93 @@ def _prepare_artifact(
...
+                    - name:version
+                    - name:alias
+                    - digest.
+        """
+        r = self._run_obj
+        assert r is not None
+        artifact = self.use_artifact(artifact_or_name=model_name)
+        assert artifact.type.lower() == "model"
+        local_path = artifact.download()
+        return local_path
```
Comment (szymon-piechowicz-wandb) [inline wandb/sdk/wandb_run.py:3127]: Nit: `use_artifact()` returns an `Artifact`, but `use_model()` returns a directory path. This is confusing.

**Review by szymon-piechowicz-wandb — COMMENTED** [review]
In the following code:
```
wandb.run.log_model("model.pt")
model = wandb.run.use_model("model")
```
I log a file but get back a directory containing that file. I'm not sure how to handle this better though...

**Comment by ibindlish** [comment]
> In the following code:
> 
> ```
> wandb.run.log_model("model.pt")
> model = wandb.run.use_model("model")
> ```
> 
> I log a file but get back a directory containing that file. I'm not sure how to handle this better though...

What if i returned the path to the file instead, if the directory has only 1 file? Otherwise, return the directory path like we do now

File: wandb/sdk/wandb_run.py [hunk 5]
```diff
@@ -3069,6 +3072,96 @@ def _prepare_artifact(
...
+            model_name: (str) - the name of the model artifact that files in 'path' will be logged to.
+            aliases: (List[str], optional) - alias(es) that will only be applied on this linked artifact
+                inside the registered model.
+                The alias "latest" will always be applied to the latest version of an artifact that is linked.
+
+        Returns:
+            None
+        """
+        name_parts = linked_model_name.split("/")
+        assert len(name_parts) == 1
```
Comment (dmitryduev) [inline wandb/sdk/wandb_run.py:3156]: I think this should be an actionable error instead. Also, it's not clear from the docstring, what it should be -- please update that and maybe even provide an example.

File: wandb/sdk/wandb_run.py [hunk 6]
```diff
@@ -3069,6 +3072,96 @@ def _prepare_artifact(
...
+        project = "model-registry"
+        target_path = self.entity + "/" + project + "/" + linked_model_name
```
Comment (dmitryduev) [inline wandb/sdk/wandb_run.py:3158]: what if we're in offline mode and entity is therefore ''? some actionable check might be quite nice.

**Review by dmitryduev — CHANGES_REQUESTED** [review]
LGTM

File: wandb/__init__.py [hunk 7]
```diff
@@ -47,6 +47,9 @@
 AlertLevel = wandb_sdk.AlertLevel
 Settings = wandb_sdk.Settings
 Config = wandb_sdk.Config
+log_model = wandb_sdk.wandb_run.Run.log_model
+use_model = wandb_sdk.wandb_run.Run.use_model
+link_model = wandb_sdk.wandb_run.Run.link_model
```
Comment (dmitryduev) [inline wandb/__init__.py:52]: You're re-defining these below
```suggestion
```

File: tests/pytest_tests/system_tests/test_core/test_model_workflow.py [hunk 8]
```diff
@@ -11,3 +14,78 @@ def test_offline_link_artifact(wandb_init):
     with pytest.raises(NotImplementedError):
         run.link_artifact(FakeArtifact(), "entity/project/portfolio", "latest")
     run.finish()
+
+
+@pytest.mark.nexus_failure(feature="models")
+def test_log_model(relay_server, wandb_init):
+    with relay_server():
```
Comment (dmitryduev) [inline tests/pytest_tests/system_tests/test_core/test_model_workflow.py:21]: doesn't look like you're using the relay server.
```suggestion
```

File: tests/pytest_tests/system_tests/test_core/test_model_workflow.py [hunk 9]
```diff
@@ -11,3 +14,78 @@ def test_offline_link_artifact(wandb_init):
...
+        run = wandb_init()
+        download_path = run.use_model("test-model:v0")
+        file = download_path
+        assert file == f"{env.get_artifact_dir()}/test-model:v0/boom.txt"
+        run.finish()
+
+
+@pytest.mark.nexus_failure(feature="models")
+def test_use_model(relay_server, wandb_init):
+    with relay_server():
```
Comment (dmitryduev) [inline tests/pytest_tests/system_tests/test_core/test_model_workflow.py:40]: same

File: tests/pytest_tests/system_tests/test_core/test_model_workflow.py [hunk 10]
```diff
@@ -11,3 +14,78 @@ def test_offline_link_artifact(wandb_init):
...
+        run = wandb_init()
+        with tempfile.TemporaryDirectory(dir="./") as tmpdir:
```
Comment (dmitryduev) [inline tests/pytest_tests/system_tests/test_core/test_model_workflow.py:42]: why does it have to be a local temp dir? can you use the standard `tmp_path` fixture instead (https://docs.pytest.org/en/6.2.x/tmpdir.html)?

File: tests/pytest_tests/system_tests/test_core/test_model_workflow.py [hunk 11]
```diff
@@ -11,3 +14,78 @@ def test_offline_link_artifact(wandb_init):
...
+            logged_artifact.wait()
+            download_path = run.use_model("test-model:v0")
+            file = download_path
+            assert file == f"{env.get_artifact_dir()}/test-model:v0/boom.txt"
+        run.finish()
+
+
+@pytest.mark.nexus_failure(feature="models")
+def test_use_model_error_artifact_type(relay_server, wandb_init):
+    with relay_server():
```
Comment (dmitryduev) [inline tests/pytest_tests/system_tests/test_core/test_model_workflow.py:59]: same

File: tests/pytest_tests/system_tests/test_core/test_model_workflow.py [hunk 12]
```diff
@@ -11,3 +14,78 @@ def test_offline_link_artifact(wandb_init):
...
+            )
+            logged_artifact.wait()
+            with pytest.raises(AssertionError):
+                _ = run.use_model("test-model:v0")
+        run.finish()
+
+
+@pytest.mark.nexus_failure(feature="models")
+def test_link_model(relay_server, wandb_init):
+    with relay_server():
```
Comment (dmitryduev) [inline tests/pytest_tests/system_tests/test_core/test_model_workflow.py:77]: same

File: wandb/sdk/wandb_run.py (hunk 6 above)
Comment (ibindlish) [reply wandb/sdk/wandb_run.py:3158]: Wouldn't `run.link_artifact()` take care of that? Although it seems like there is no offline-model implementation for link artifact 😅

Comment (ibindlish) [reply wandb/sdk/wandb_run.py:3179]: @dmitryduev Actually, I think i don't understand what actionable check we would add here. I tried running this in offline mode, and the `entity` didn't get set to `''`. 

Even if it did, the path that gets passed to `link_artifact()` would be `/model-registry/<linked_model_name>`, which doesn't cause any issues as the interface _publish method assigns entity from the settings too. Am i missing something?

**Comment by noaleetz** [comment]
Hey @ibindlish - who are we still waiting on a review on here?

**Comment by ibindlish** [comment]
> Hey @ibindlish - who are we still waiting on a review on here?

@noaleetz just trying to get this comment resolved: https://github.com/wandb/wandb/pull/6205#discussion_r1346269413

**Review by dmitryduev — APPROVED** [review]

**Comment by noaleetz** [comment]
Note to self - will be released with next SDK release - release version: sdk-0.11.1
